jormungandr/src/fragment/pool.rs

```diff
@@ -47,56 +51,99 @@ impl Pools {
         &mut self.logs
     }
 
-    /// Returns number of registered fragments
+    /// Returns number of registered fragments. Setting `fail_fast` to `true` will force this
+    /// method to reject all fragments after the first invalid fragments was met.
     pub async fn insert_and_propagate_all(
         &mut self,
         origin: FragmentOrigin,
-        mut fragments: Vec<Fragment>,
-    ) -> Result<usize, Error> {
+        fragments: Vec<Fragment>,
+        fail_fast: bool,
+    ) -> Result<FragmentsProcessingSummary, Error> {
         tracing::debug!(origin = ?origin, "received {} fragments", fragments.len());
-        fragments.retain(is_fragment_valid);
-        if fragments.is_empty() {
-            tracing::debug!("none of the received fragments are valid");
-            return Ok(0);
-        }
+
         let mut network_msg_box = self.network_msg_box.clone();
-        let fragment_ids = fragments.iter().map(Fragment::id).collect::<Vec<_>>();
-        let fragments_exist_in_logs = self.logs.exist_all(fragment_ids);
-        let new_fragments = fragments
-            .into_iter()
-            .zip(fragments_exist_in_logs)
-            .filter(|(_, exists_in_logs)| !exists_in_logs)
-            .map(|(fragment, _)| fragment);
-
-        let mut max_added = 0;
-
-        for (i, pool) in self.pools.iter_mut().enumerate() {
-            let new_fragments = pool.insert_all(new_fragments.clone());
+
+        let mut filtered_fragments = Vec::new();
+        let mut rejected = Vec::new();
+
+        let mut fragments = fragments.into_iter();
+
+        for fragment in fragments.by_ref() {
+            let id = fragment.id();
+
+            if self.logs.exists(id) {
+                rejected.push(RejectedFragmentInfo {
+                    id,
+                    reason: FragmentRejectionReason::FragmentAlreadyInLog,
+                });
+                continue;
+            }
+
+            if !is_fragment_valid(&fragment) {
+                rejected.push(RejectedFragmentInfo {
+                    id,
+                    reason: FragmentRejectionReason::FragmentInvalid,
+                });
+
+                if fail_fast {
+                    break;
+                }
+
+                continue;
+            }
+
+            filtered_fragments.push(fragment);
+        }
+
+        if fail_fast {
+            for fragment in fragments {
+                rejected.push(RejectedFragmentInfo {
+                    id: fragment.id(),
+                    reason: FragmentRejectionReason::PreviousFragmentInvalid,
+                })
+            }
+        }
```

Comment:
Are you sure we want to do this as the general case?

Maybe a better approach for client-submitted fragments would be to only submit and log them if all of them validate (the function seems stateless between fragments), otherwise a HTTP failure code will signal atomic rejection of all of them?

Reply:
Thinking aloud: `/v1/fragments` might be used by a client to submit some fragments that make sense together (even if they should not be dependent on one another), so the possibility that only part of them may be accepted would bring complication into client-side error handling.

Reply:
OTOH, we do want the ability to spill fragments that don't fit into the mempool, so would the mempool fitting logic need to apply to the whole batch to make things consistent?

Reply:
Anyway, any of the fragments may end up not included into a confirmed block, so partial rejection is OK in principle. It's just that if a client sent something so wrong in one of the fragments, we should not assume any of their fragments make sense, so no need for such elaborate partial acceptance logic and weird error reasons like "it was the fault in another fragment that made this one fail".